pkg/agent/plugin/workloadattestor/k8s/sigstore/sigstore.go

```diff
@@ -0,0 +1,486 @@
+	}
+	if skip {
+		return []string{signatureVerifiedSelector}, nil
+	}
+
+	imageID := status.ImageID
+
+	cachedSignature := s.sigstorecache.GetSignature(imageID)
+	if cachedSignature != nil {
+		s.logger.Debug("Found cached signature", "image_id", imageID)
+	} else {
+		signatures, err := s.FetchImageSignatures(ctx, imageID)
+		if err != nil {
+			return nil, err
+		}
+
+		selectors := s.ExtractSelectorsFromSignatures(signatures, status.ContainerID)
+
+		cachedSignature = &Item{
+			Key:   imageID,
+			Value: selectors,
+		}
+
+		s.logger.Debug("Caching signature", "image_id", imageID)
+		s.sigstorecache.PutSignature(*cachedSignature)
+	}
+
+	var selectorsString []string
+	if len(cachedSignature.Value) > 0 {
+		for _, selector := range cachedSignature.Value {
+			toString := selectorsToString(selector, status.ContainerID)
+			selectorsString = append(selectorsString, toString...)
+		}
+		selectorsString = append(selectorsString, signatureVerifiedSelector)
+	}
+
+	return selectorsString, nil
+}
+
+func (s *sigstoreImpl) SetRekorURL(rekorURL string) error {
+	if rekorURL == "" {
+		return errors.New("rekor URL is empty")
+	}
+	rekorURI, err := url.Parse(rekorURL)
+	if err != nil {
+		return fmt.Errorf("failed parsing rekor URI: %w", err)
+	}
+	if rekorURI.Host == "" {
+		return fmt.Errorf("host is required on rekor URL")
+	}
+	if rekorURI.Scheme != "https" {
+		return fmt.Errorf("invalid rekor URL Scheme %q", rekorURI.Scheme)
+	}
+	s.rekorURL = *rekorURI
+	return nil
+}
+
+func getSignatureSubject(signature oci.Signature) (string, error) {
+	if signature == nil {
+		return "", errors.New("signature is nil")
+	}
+	ss := payload.SimpleContainerImage{}
+	pl, err := signature.Payload()
+	if err != nil {
+		return "", err
+	}
+	if pl == nil {
+		return "", errors.New("signature payload is nil")
+	}
+	if err := json.Unmarshal(pl, &ss); err != nil {
+		return "", err
+	}
+	cert, err := signature.Cert()
+	if err != nil {
+		return "", fmt.Errorf("failed to access signature certificate: %w", err)
+	}
+
+	if cert != nil {
+		return sig.CertSubject(cert), nil
+	}
+	if len(ss.Optional) > 0 {
+		if subjString, ok := ss.Optional["subject"]; ok {
+			if subj, ok := subjString.(string); ok {
+				return subj, nil
+			}
+		}
+	}
+
+	return "", errors.New("no subject found in signature")
+}
+
+func getSignatureProvider(signature oci.Signature) (string, error) {
+	if signature == nil {
+		return "", errors.New("signature is nil")
+	}
+	cert, err := signature.Cert()
+	if err != nil {
+		return "", fmt.Errorf("failed to access signature certificate: %w", err)
+	}
+	if cert == nil {
+		return "", errors.New("no certificate found in signature")
+	}
+	return certOIDCProvider(cert)
+}
+
+func certOIDCProvider(cert *x509.Certificate) (string, error) {
+	if cert == nil {
+		return "", errors.New("certificate is nil")
+	}
+	// OIDC token issuer Object Identifier
+	objectIdentifier := asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 57264, 1, 1}
```

Comment:
This is the OID for the OIDC token issuer, it's a registration number for certificate extensions (in this case, sigstore owns OID "1, 3, 6, 1, 4, 1, 57264" and its child OIDs, in this case the OIDC Token Issuer), which we need to retrieve the issuer for the allowlist. In any case, this will be moved to a constant.